solidity/contracts/ECDSARewardsDistributor.sol

```diff
@@ -0,0 +1,134 @@
+/// @dev This contract is based on the Uniswap's Merkle Distributor
+/// https://github.com/Uniswap/merkle-distributor with some modifications:
+/// - added a map of merkle root keys with the amount of KEEP (value) that will
+///   be allocated for those merkle roots
+/// - added receiveApproval() function that will be called each time to allocate
+///   new KEEP rewards for a given merkle root. Merkle root is going to be generated
+///   regulary (ex. every week) and it is also means that an interval for that
+///   merkle root has passed
+/// - changed code accordingly to process claimed rewards using a map of merkle
+///   roots
+contract ECDSARewardsDistributor is Ownable {
+    using SafeERC20 for KeepToken;
+    using BytesLib for bytes;
+    using SafeMath for uint256;
+
+    KeepToken public token;
+
+    // This event is triggered whenever a call to #claim succeeds.
+    event RewardsClaimed(uint256 index, address account, uint256 amount);
+    // This event is triggered whenever rewards are allocated.
+    event RewardsAllocated(bytes32 merkleRoot, uint256 amount);
+
+    // Merkle root -> total amount for distribution for a given timeframe.
+    mapping(bytes32 => uint256) private merkleRoots;
+    // Bytes32 key is a merkle root and the value is a packed array of booleans.
+    mapping(bytes32 => mapping(uint256 => uint256)) private claimedBitMap;
+
+    constructor(address _token) public {
+        token = KeepToken(_token);
+    }
+
+    function claim(
+        uint256 index,
+        address account,
+        uint256 amount,
+        bytes32 merkleRoot,
+        bytes32[] calldata merkleProof
+    ) external {
+        require(!isClaimed(index, merkleRoot), "Reward already claimed");
+        require(
+            merkleRoots[merkleRoot] > 0,
+            "Rewards must be allocated for a given merkle root"
+        );
+
+        // Verify the merkle proof.
+        bytes32 node = keccak256(abi.encodePacked(index, account, amount));
+
+        require(
+            MerkleProof.verify(merkleProof, merkleRoot, node),
+            "Invalid proof"
+        );
+
+        // Mark it claimed and send the token.
+        _setClaimed(index, merkleRoot);
+        require(IERC20(token).transfer(account, amount), "Transfer failed");
+
+        // Update KEEP amount for the given merkleRoot
+        merkleRoots[merkleRoot] = merkleRoots[merkleRoot].sub(amount);
+
+        emit RewardsClaimed(index, account, amount);
+    }
+
+    /// Call receiveApproval to allocate amount of KEEP for a given merkle root.
+    /// @param _from The original sender of the tokens.
+    /// @param _amount The amount of KEEP tokens to fund.
+    /// @param _token The KEEP token to fund the rewards in.
+    /// @param _extraData Merkle root (32 bytes).
+    function receiveApproval(
+        address _from,
+        uint256 _amount,
+        address _token,
+        bytes memory _extraData
+    ) public onlyOwner {
```

Comment:
To avoid `onlyOwner` on `receiveApproval` (it can be problematic because for ERC20 `token.approveAndCall`, `token` is the `msg.sender` in `receiveApproval`), we can add a separate function `function allocate(bytes32 merkleRoot, uint256 amount) public onlyOwner` that will call `token.safeTransferFrom(msg.sender, address(this), amount);`. Then, in the contract calling this function, we'll do:
```
token.approve(address(distributor), amount);
distributor.allocate(merkleRoot, amount);
```

Reply:
Yes, it is problematic and I experienced that in tests. Changed in 
20f7d19